src/parameterizations/lateral/MOM_Zanna_Bolton.F90

```diff
@@ -173,726 +193,802 @@ subroutine ZB_2020_init(Time, GV, US, param_file, diag, CS, use_ZB2020)
+  integer :: niter                       ! required number of iterations
 
-  call compute_energy_source(u, v, h, fx, fy, G, GV, CS)
+  integer :: is, ie, js, je, Isq, Ieq, Jsq, Jeq, nz
+  integer :: i, j, k, n
 
-end subroutine Zanna_Bolton_2020
+  niter = CS%HPF_iter
 
-!> Filter which is used to smooth velocity gradient tensor
-!! or the stress tensor.
-!! If n_lowpass and n_highpass are positive,
-!! the filter is given by:
-!! I - (I-G^n_lowpass)^n_highpass
-!! where I is the identity matrix and G is smooth_Tq().
-!! It is filter of order 2*n_highpass,
-!! where n_lowpass is the number of iterations
-!! which defines the filter scale.
-!! If n_lowpass is negative, returns residual
-!! for the same filter:
-!! (I-G^|n_lowpass|)^n_highpass
-!! Input does not require halo. Output has full halo.
-subroutine filter(G, mask_T, mask_q, n_lowpass, n_highpass, T, q)
-  type(ocean_grid_type), intent(in) :: G !< Ocean grid
-  integer, intent(in) :: n_lowpass  !< number of low-pass iterations
-  integer, intent(in) :: n_highpass !< number of high-pass iterations
-  real, dimension(SZI_(G),SZJ_(G)),   &
-                          intent(in) :: mask_T !< mask of wet points in T (CENTER) points [nondim]
-  real, dimension(SZIB_(G),SZJB_(G)), &
-                          intent(in) :: mask_q !< mask of wet points in q (CORNER) points [nondim]
-  real, dimension(SZI_(G),SZJ_(G)),   &
-             optional, intent(inout) :: T      !< any field at T (CENTER) points [arbitrary]
-  real, dimension(SZIB_(G),SZJB_(G)), &
-             optional, intent(inout) :: q      !< any field at q (CORNER) points [arbitrary]
+  if (niter == 0) return
 
-  real, dimension(SZIB_(G),SZJB_(G)) :: q1, q2          ! intermediate q-fields [arbitrary]
-  real, dimension(SZI_(G),SZJ_(G))   :: T1, T2          ! intermediate T-fields [arbitrary]
-  real :: max_before, min_before, max_after, min_after  ! minimum and maximum values of fields
-                                                        ! before and after filtering [arbitrary]
+  is  = G%isc  ; ie  = G%iec  ; js  = G%jsc  ; je  = G%jec ; nz = GV%ke
+  Isq = G%IscB ; Ieq = G%IecB ; Jsq = G%JscB ; Jeq = G%JecB
 
-  integer :: i_highpass, i_lowpass
-  integer :: i, j
-  integer :: is, ie, js, je, Isq, Ieq, Jsq, Jeq
+  if (.not. G%symmetric) &
+    call do_group_pass(CS%pass_xx, G%Domain, &
+      clock=CS%id_clock_mpi)
 
-  is  = G%isc  ; ie  = G%iec  ; js  = G%jsc  ; je  = G%jec
-  Isq = G%IscB ; Ieq = G%IecB ; Jsq = G%JscB ; Jeq = G%JecB
+  ! This is just copy of the array
+  call cpu_clock_begin(CS%id_clock_filter)
+    do k=1,nz
+      ! Halo of size 2 is valid
+      do j=js-2,je+2; do i=is-2,ie+2
+        sh_xx(i,j,k) = CS%sh_xx(i,j,k)
+      enddo; enddo
+      ! Only halo of size 1 is valid
+      do J=Jsq-1,Jeq+1; do I=Isq-1,Ieq+1
+        sh_xy(I,J,k) = CS%sh_xy(I,J,k)
+        vort_xy(I,J,k) = CS%vort_xy(I,J,k)
+      enddo; enddo
+    enddo
+  call cpu_clock_end(CS%id_clock_filter)
```

Comment:
These copies are costly. Apparently, they cost the same as 4 iterations of the filter. If the updated halo size allows me to do all filter iterations in a single matching halo step, I can copy just a 2D slice instead of the full 3D array.

However, I postpone these optimizations for future. Note that the principal possibility of using this optimization depends on the number of filter iterations and/or halo size in this module which can be increased for optimization reasons.